velox/vector/VectorBatch.h

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "velox/buffer/Buffer.h"
+#include "velox/common/memory/ByteStream.h"
+#include "velox/common/memory/MappedMemory.h"
+#include "velox/common/memory/Memory.h"
+#include "velox/common/memory/StreamArena.h"
+#include "velox/type/Type.h"
+#include "velox/vector/SelectivityVector.h"
+
+namespace facebook::velox {
+
+class RowVector;
+
+namespace batch {
+
+enum class BatchSerdeStatus {
+  Success,
+  AllocationFailed,
+  SerializationFailed,
+  DeserializationFailed,
+  KeysSerializationFailed,
+  SizesDeserializationFailed
+};
+
+class VectorKeySerializer {
+ public:
+  /// Deserializes a set of keys in a vector an aggregated key
+  /// \param vector the input row vector
+  /// \param keys the indexes of the keys in the row vector
+  /// \param row the row index
+  /// \param serializedKeysBuffer a pre-allocated block for serialization
+  /// \param serializedKeys the return serialized block
+  /// \return the size of the serialized keys in the input block
+  virtual std::optional<size_t> serializeKeys(
+      const std::shared_ptr<RowVector>& vector,
+      const std::vector<int>& keys,
+      vector_size_t row,
+      std::string_view& serializedKeysBuffer,
+      std::string_view& serializedKeys) = 0;
+  virtual ~VectorKeySerializer() = default;
+};
+
+class VectorSerde {
+ public:
+  VectorSerde(const std::unique_ptr<VectorKeySerializer>& keySerializer)
+      : keySerializer_{std::move(keySerializer)} {}
+  virtual ~VectorSerde() = default;
+
+  /// Must be called every time a new vector will be processed
+  /// \param vector the new vector to be serialized
+  virtual void initialize(const std::shared_ptr<RowVector>& vector) = 0;
```

Comment:
Cant this be moved to constructor and some static factory cretaed for the Serde so people do not have to remember this kind of state ?

Reply:
No, the serializer can go through multiple vectors.

Reply:
You remember this takes care of pre-loading of the input vector and after that we can use the size calculator safely. Yes, the user is supposed to call initialize before serializing any vector. The pattern in which it works is like this:
```<constructor> <initialize> <serialize row x> ... <serialize row y> <initialize> <serialize row xx> ... <serialize row yy> ...```
I can try to change it so each time we need to ```initialize``` we create a new object but all the methods there are static.

Reply:
I will call it ```reset```